File: Exercises/Regression/simple_linear_regression.R
# Regresi??n Lineal Simple

#----------PREPROCESAMIENTO DE DATOS-------------------
#Importar dataset
#dataset = read.csv("..\datasets\Part 1 - Data Preprocessing\Section 2 -------------------- Part 1 - Data Preprocessing --------------------\Data.csv")
dataset = read.csv("Salary_Data.csv")
dataset = dataset[,1:3]


##Dividir en entrenamiento y Test

#install.packages("caTools")
library(caTools)

set.seed(123)
split = sample.split(dataset$Salary,SplitRatio = 2/3)
training_set = subset(dataset,split==TRUE)
testing_set = subset(dataset,split==FALSE)


#Escalado de valores
# training_set[,2:3] = scale(training_set[,2:3])
# testing_set[,2:3] = scale(testing_set[,2:3])


#----------CONSTRUCCI??N DEL MODELO -------------------

#Ajustar modelo de regresi??n lineal simple - Datos entrenamiento
regressor = lm(formula = Salary ~ YearsExperience, training_set)

#Predecir los resultados con el conjunto de test
y_pred = predict(regressor,newdata = testing_set)


#-----------VISUALIZACI??N DE RESULTADOS---------------
#install.packages("ggplot2")
library("ggplot2") 

#Visualizaci??n de conjunto de datos de entrenamiento
ggplot() +
  geom_point(aes(x=training_set$YearsExperience, y = training_set$Salary),
             colour="red") + 
  geom_line(aes(x=training_set$YearsExperience, 
                y=predict(regressor,newdata = training_set)),
            colour = "blue") +
  ggtitle("Sueldo vs. A??os de experiencia (Conjunto de entrenamiento)") + 
  xlab("A??os de experiencia") + 
  ylab("Sueldo (en $)")
  

#Visualizaci??n de conjunto de datos de prueba
ggplot() +
  geom_point(aes(x=testing_set$YearsExperience, y = testing_set$Salary),
             colour="red") + 
  geom_line(aes(x=training_set$YearsExperience, 
                y=predict(regressor,newdata = training_set)),
            colour = "blue") +
  ggtitle("Sueldo vs. A??os de experiencia (Conjunto de pruebas)") + 
  xlab("A??os de experiencia") + 
  ylab("Sueldo (en $)")

File: Exercises/Regression/simple_linear_regression.py
# -*- coding: utf-8 -*-

# Regresión Lineal Simple

#Importar librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#------------------PREPROCESAMIENTO DE DATOS---------------------

#Cargar el Dataset
dataset = pd.read_csv("../../datasets/Part 2 - Regression/Section 4 - Simple Linear Regression/Salary_Data.csv")
x = dataset.iloc[:,:-1].values
y = dataset.iloc[:,1].values

#Tratamiento de nulos
from sklearn.preprocessing import Imputer
imputer = Imputer(missing_values="NaN", strategy="mean",axis=0)
imputer = imputer.fit(x[:,:])
x[:,:] = imputer.transform(x[:,:])

#Dividir Datastet en entrebamiento y pruebas
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=1/3,random_state=0)

#------------------CONSTRUCCIÓN DEL MODELO---------------------

#Modelo de Regresión Lineal Simple - Datos Entrenamiento
from sklearn.linear_model import LinearRegression
regression = LinearRegression()
regression.fit(x_train,y_train)

#Predecir el conjunto Test
y_pred = regression.predict(x_test)

#------------------VISUALIZAR LOS RESULTADOS---------------------
#Visualizar datos de entrenamiento
plt.scatter(x_train,y_train, color="red")
plt.plot(x_train,regression.predict(x_train), color="blue")
plt.title("Sueldo vs. Años de experiencia (Conjunto de entrenamiento")
plt.xlabel("Años de experiencia")
plt.ylabel("Salario($)")
plt.show()

#Visualizar datos de test
plt.scatter(x_test,y_test, color="red")
plt.plot(x_train,regression.predict(x_train), color="blue")
plt.title("Sueldo vs. Años de experiencia (Conjunto de prueba")
plt.xlabel("Años de experiencia")
plt.ylabel("Salario($)")
plt.show()